docs/reference/c++_support.rst

```diff
@@ -0,0 +1,165 @@
+binary. A managed variable can be accessed in both device and host code. It is important
+to know where each variable is because they will be only available in certain places.
+Generally, variables allocated in the host memory will not be available from device code
+and variables allocated in the device memory will not be available from host code. This
+can be an issue mostly with pointers in host memory, which are pointing to device memory.
+Dereferencing these will cause a segmentation fault.
+
+Exception handling
+-------------------------------------------------------------------------------
+
+An important difference between the host and device code is exception handling. In device
+code it is not available, error handling needs to be done with return codes. This is
+because of the hardware architecture does not allow for this kind of control flow.
+
+Kernel parameters
+-------------------------------------------------------------------------------
+
+There are some restrictions on kernel function parameters. They cannot be passed by
+reference, as these functions run on the device, but are called from the host. Also,
+variable number of arguments is not allowed.
+
+Classes
+-------------------------------------------------------------------------------
+
+Classes work on both host and device side, but there are some constraints. ``static``
+data members need to be ``const`` qualified, and ``static`` member functions can't be
+``__global__``. ``Virtual`` member functions work, but it's undefined behaviour to call a
+virtual function from the host, when the object was created on the device, or the
+other way around. This also means, that you can't pass an object with virtual functions
+as a parameter to a kernel.
+
+Polymorphic Function Wrappers
+-------------------------------------------------------------------------------
+
+Since C++11 the standard library has a polymorphic function wrapper ``std::function``.
+This has an equivalent in CUDA called ``nvstd::function`` to work on CUDA enabled
+devices. Unfortunately, HIP doesn't have its own version currently.
+
+Extended Lambdas
+-------------------------------------------------------------------------------
+
+Lambdas are a powerful tool in modern C++ and are supported by the HIP ecosystem. By
+default, the lambda will work as you expect, but keep in mind that they will inherit the
+execution space specification from the surrounding context. For example, in a device
+the lambda can only be called from other device functions. This also means that lambdas
+can't be used as template argument for kernels unless they are defined in a device
+function or a kernel. To help develop versatile software, HIP has an extension making
+lambdas even more powerful. They can have ``__host__`` or ``__device__`` qualifiers. This
+way developers can define lambdas in host code, that can run on the device side as well,
+and used as template parameter for ``__global__`` functions.
```

Comment:
HIP supports Lambdas, which by default work as expected.

Lambdas inherit the execution space specification from the surrounding context. For example, in a device, the lambda can only be called from other device functions. This also means that lambdas can’t be used as a template argument for kernels unless they are defined in a device function or a kernel.

To help develop versatile software, HIP supports an extension that makes lambdas even more powerful. They can have ``__host__`` or ``__device__`` qualifiers. Developers can use this feature to define lambdas in host code that can run on the device side and be used as a template parameter for ``__global__`` functions.